### DISCUSSION TITLE
Click & drag select, copy-paste

### DISCUSSION THREAD

**Discussion by guiv42** [helge17/tuxguitar#29] [General]
Hi
I'm currently considering to merge somehow the awesome "Click & drag" select feature from 2.0beta branch. If someone could play a bit with it, I would appreciate a feedback on the user interface.
My (very personal) point of view: I don't see the added value to have 2 independent features "copy / paste" and "copy measure / paste measure". These features are accessible through independent menu items, which may be enabled or disabled depending of context. Typically, "copy" is only enabled if a selection is active.
E.g. the following sequence seems counter-intuitive to me : select something with the mouse, copy, unselect, copy measure. There, "paste" action is enabled but doesn't paste anything.

My proposal:
- delete "copy measure / paste measure" from menu
- only keep "copy / paste" in menu "Edit", associated with default shortcuts Ctrl-C, Ctrl-V
- When copying:
-- if no selection is active, open existing "copy measure" dialog
-- else, copy selection to clipboard
- When pasting:
-- if clipboard contains measure(s), open existing "paste measure" dialog
-- else (selection in clipboard): paste selection

Any comments welcome, user interface is a very subjective area

**Comment by helge17** [comment]
Your proposal sounds good to me. Since you can Shift-click to select individual note ranges in 2.0beta, it might be great to be able to Ctrl-click to select whole measures.

I can also create builds from your fork for testing if that helps.

**Comment by guiv42** [comment]
After a few more tests:
1. There is a significant difference between "paste selection" and "paste measure" in 2.0
"Paste measure" dialog enables user to choose paste behavior, either inserting the pasted notes (shifts everything right), or just replacing the existing content.
I don't understand the code of "paste selection" feature on that point: the 2 options are handled... but I couldn't link this to user interface. How could a user choose between the 2 options, since no dialog opens when pasting?
3. What should be the behavior in the following sequence?
- select something with the mouse, copy
- select something *else*, paste
Where to paste? Currently 2.0 pastes at the *caret* place. This can be a little confusing: depending of the direction of mouse movement in the second selection (left-to-right or right-to-left), it will not paste at the same place.
Anyway I don't know if there is a good answer to that question. If I select 2 notes, copy, then select 4 measures and paste... Where would I expect the 2 notes to be pasted??

**Comment by guiv42** [comment]
Hi,
there is a first version including the "click & drag select" feature in a [dedicated branch of my fork](https://github.com/guiv42/tuxguitar/tree/selector). It's definitely not mature enough for a merge, but probably enough to start some testing.
What is supposed to work:
- select notes with click & drag, extend selection with shift+click
- delete selection (with 'del' key)
- cut / copy / paste: accessible through "Edit" menu
- both SWT and JFX configs should work
- cut/paste/delete should be undoable
- Ctrl-c and -v shortcuts are reallocated to copy/paste. This might require you to delete your tuxguitar conf file, or in Tools/Shortcuts to restore to defaults
- Ctrl-x shortcut is allocated to "Cut"
- selection color can be customized in Tools/Settings

As mentioned above:
- when copying, if something is selected it is copied, else "copy measure" dialog is opened
- similarly, Paste feature either pastes copied selection, or opens "paste measure" dialog depending from clipboard's content
- cut feature is only enabled if something is selected

Android version has just been adapted to be compliant with all modifications. It should build, and work as before (no new feature included).

Many other features are possible in 2.0b branch with a selection, I prefer to go step-by-step and not merge everything all together. This is already a very significant evolution.

If anyone has some time to test this, I'll appreciate your feedback!

**Comment by guiv42** [comment]
builds **for testing purposes** are available [here](https://github.com/guiv42/tuxguitar/releases)
only Linux & Windows, only SWT

**Comment by helge17** [comment]
I quickly tested your selector builds. Both Linux and Windows already look pretty good!

Two comments:

1. "Edit - Copy" OK
"Edit - Paste"   OK
"Edit - Undo"    OK
"Edit - Redo"    does "something different" than Paste

2. "Edit - Paste" should ask something like "insert or overwrite?", like "Measure - Paste Measure" does.

**Reply by guiv42** [reply]
Thank you very much for this feedback. I'll try to spend some time on testing also. Nice to see the win version works correctly, thanks for [install.md](https://github.com/helge17/tuxguitar/blob/master/INSTALL.md), it really helped.
1. "redo" KO
I found the root cause, fix is coming soon

2. overwrite vs. insert: this is a more complex area... user interface is really a challenge

Already spent a significant amount of time on this topic ("2.0beta" is definitely not perfect). In 2.0b, the code for "Paste selection" feature *can theoretically* either overwrite or insert... but I never found the associated ui. So the feature is easily accessible, but what could be the best user interface? I don't know.
In a first step I opted for a simple behavior:
- always overwrite when pasting a selection (like 2.0b does, as far as I could test)
- I added "re-select content after pasting a selection" feature: I think it has a real added value, showing clearly what has been pasted and where. If overwriting was not appropriate at least it is visible, and Undo is accessible

 I think opening a dialog at each copy/paste action is not very user-friendly (never seen any app behaving like this). Possibly add a user setting to configure "overwrite / insert"? Any idea welcome!

**Comment by guiv42** [comment]
Following merge of this feature in master (e12be785182856a96059bfd90579d5678d6c0909) I close this discussion.
Please use issues to request an evolution
